src/Pagination/Connection.hack

```diff
@@ -90,17 +89,49 @@ abstract class Connection {
      * Whether a page exists before the `after` cursor, was such a cursor provided.
      * Can be overridden for efficiency.
      */
-    public async function hasNextPage(string $after_cursor): Awaitable<bool> {
-		return C\count(await $this->fetch(shape('first' => 1, 'after' => $after_cursor))) > 0;
-	}
+    protected async function hasNextPage(
+        PaginationArgs $args,
+        string $start_cursor,
+        string $end_cursor,
+    ): Awaitable<bool> {
+        if (Shapes::keyExists($args, 'last') && !Shapes::keyExists($args, 'before')) {
+            return false;
+        }
+
+        if (
+            Shapes::keyExists($args, 'after') &&
```

Comment:
Why does it matter if `after` is provided here? Isn't it the same if we start at the first cursor?